Proposed syntax APIs

ilya-biryukov opened the pull request.

ilya-biryukov committed:
[Syntax] Introduce TokenBuffer, start clangToolingSyntax library

Summary:
TokenBuffer stores the list of tokens for a file obtained after
preprocessing. This is a base building block for syntax trees,
see [1] for the full proposal on syntax trees.

This commits also starts a new sub-library of ClangTooling, which
would be the home for the syntax trees and syntax-tree-based refactoring
utilities.

[1]: https://lists.llvm.org/pipermail/cfe-dev/2019-February/061414.html

Reviewers: gribozavr

Subscribers: mgorny, jdoerfert, cfe-commits

Tags: #clang

Differential Revision: https://reviews.llvm.org/D59887

ilya-biryukov committed:
[Lex] Allow to consume tokens while preprocessing

Summary:
By adding a hook to consume all tokens produced by the preprocessor.
The intention of this change is to make it possible to consume the
expanded tokens without re-runnig the preprocessor with minimal changes
to the preprocessor and minimal performance penalty when preprocessing
without recording the tokens.

The added hook is very low-level and reconstructing the expanded token
stream requires more work in the client code, the actual algorithm to
collect the tokens using this hook can be found in the follow-up change.

Reviewers: rsmith

Subscribers: cfe-commits

Tags: #clang

Differential Revision: https://reviews.llvm.org/D59885

ilya-biryukov committed:
[AST] Make CXXBindTemporaryExpr final. NFC

ilya-biryukov committed:
[AST] Treat inherited default argument as implicit code

When traversing with RecursiveASTVisitor.

gribozavr commented on clang/include/clang/AST/ExprCXX.h:
```diff
@@ -1213,7 +1213,7 @@ class CXXTemporary {
 ///   }
 /// \endcode
-class CXXBindTemporaryExpr : public Expr {
+class CXXBindTemporaryExpr final : public Expr {
```
You can totally submit this change today.

gribozavr commented on clang/include/clang/AST/RecursiveASTVisitor.h:
```diff
@@ -2084,6 +2084,9 @@ DEF_TRAVERSE_DECL(NonTypeTemplateParmDecl, {
 
+  if (!shouldVisitImplicitCode() && D->hasInheritedDefaultArg())
+    return true;
+
```
Same, I think you can submit this today, as long as you add a test.

gribozavr commented on clang/lib/Lex/Preprocessor.cpp:
```diff
@@ -893,6 +895,8 @@ void Preprocessor::Lex(Token &Result) {
   }
 
   LastTokenWasAt = Result.is(tok::at);
+  if (LLVM_UNLIKELY(OnToken) && (ReturnedToken && !Cached))
```
No need for extra parens on the right?

gribozavr commented on clang/include/clang/Lex/Preprocessor.h:
```diff
@@ -843,6 +845,10 @@ class Preprocessor {
+  void watchTokens(llvm::unique_function<void(const clang::Token &)> F) {
+    OnToken = std::move(F);
+  }
+
```
I think this API should also be pushed out separately, with a test of course.

gribozavr commented on clang/tools/clang-syntax/SyntaxMain.cpp:
```diff
@@ -0,0 +1,163 @@
+//===- SyntaxMain.cpp -----------------------------------------*- C++ -*-=====//
```
Do you intend to commit this file upstream, or is it just for your own testing?

gribozavr commented on clang/tools/clang-syntax/SyntaxMain.cpp:
```diff
@@ -0,0 +1,163 @@
+    OS << "\n";
+    return;
+  }
+  auto *T = llvm::cast<syntax::TreeNode>(N);
```
`auto *TN`?

gribozavr commented on clang/tools/clang-syntax/SyntaxMain.cpp:
```diff
@@ -0,0 +1,163 @@
+      OS << "|-";
+
+    printTree(OS, *It, SM, LO, Level + 1);
+  }
```
I don't see code that prints "`| `" for continuation lines.

```
        Foo
        |-Bar
here->  | |-Baz
        `-Bar
```

gribozavr commented on clang/tools/clang-syntax/SyntaxMain.cpp:
```diff
@@ -0,0 +1,163 @@
+int main(int argc, const char **argv) {
+  llvm::sys::PrintStackTraceOnErrorSignal(argv[0]);
+  auto Executor = clang::tooling::createExecutorFromCommandLineArgs(
+      argc, argv, llvm::cl::GeneralCategory,
```
How does this mash with the `llvm::cl::opt` flags?  Don't we need to remove them before passing `argv` to libTooling?

gribozavr commented on clang/tools/clang-syntax/SyntaxMain.cpp:
```diff
@@ -0,0 +1,163 @@
+      }
+
+    private:
+      syntax::TokenCollector Tokens;
```
TheTokenCollector?  `Tokens` sounds like a vector or some other container.

gribozavr commented on clang/tools/clang-syntax/Examples.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+/// Replaces a*b with a+b and vice versa. Can parenthesize.
+LLVM_NODISCARD syntax::Expr *swapMultipliesAndAdditions(syntax::Context &C,
```
"Multiplications"

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  virtual ~Node() = default;
+  virtual NodeKind kind() const = 0;
+
+  /// Construct a deep copy of the node. Useful in case the current node will be
```
"/// Returns a deep copy of the node."

How to use it is up to the caller.

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  Node *copy(Context &C) const;
+  /// Checks if the node can be copied. See isModifiable() for details,
+  /// preconditions for copy are the same.
+  bool isCopyable(const Context &C) const { return isModifiable(C); }
```
Why is `isCopyable` a separate function, if it is only an alias?

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  TreeNode *Parent = nullptr;
+};
+
+/// A leaf node, pointing to a consecutive range of tokens in some token buffer.
```
What is "token buffer"?

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  llvm::ArrayRef<Token> Tokens;
+};
+
+/// A pointer into a particular token inside a leaf node.
```
"A pointer to ..."

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  }
+
+  /// EXPECTS: Child->isModifiable() && Replacement->isCopyable()
+  Node *replaceChild(Context &C, Node *Child, const Node *Replacement);
```
Make this an API on the child node?

```
node->replaceWith(Replacement, Context);
```

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  void addChild(Node *C);
+
+protected:
+  /// Find the child that matches \p Predicate, skipping first \p Skip matches.
```
"the first child"

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  /// This is not allowed to fail, the callers are expected to know the
+  /// structure of the AST.
+  std::vector<Node *>::iterator
+  findChild(llvm::function_ref<bool(Node *)> Predicate, int Skip = 0);
```
"findFirstChild"

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+
+protected:
+  /// Find the child that matches \p Predicate, skipping first \p Skip matches.
+  /// This is not allowed to fail, the callers are expected to know the
```
"this is not allowed to fail" sounds like a spec, not a doc.

"This function never fails"?.. but then I don't understand what it means -- that it never returns the end iterator or what?

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+/// Perform a post-order traversal of the syntax tree, calling \p Visit at each
+/// node.
+void traverse(Node *N, llvm::function_ref<void(Node *)> Visit);
+void traverse(const Node *N, llvm::function_ref<void(const Node *)> Visit);
```
Move right under `class Node`?  or even make them methods of `Node`?

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+void traverse(Node *N, llvm::function_ref<void(Node *)> Visit);
+void traverse(const Node *N, llvm::function_ref<void(const Node *)> Visit);
+
+/// A tree node of an unknown kind, i.e. a syntax error or an unimplemented
```
Unclear what "unimplemented" means -- unimplemented in the compiler, or unimplemented in the syntax tree?  If the latter, what's the difference with `UnknownExpr`?

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+/// An expression in the clang AST which is missing an equivalent in the syntax
+/// tree.
+/// This will be removed when we handle all of the AST nodes.
+class UnknownExpr : public TreeNode {
```
Should derive from Expr?

ilya-biryukov replied:
Absolutely! Thanks!

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  NodeKind kind() const override { return NodeKind::LiteralExpr; }
+};
+
+class PredefinedExpr final : public Expr {
```
What's this?

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  Expr *rhs();
+};
+
+/// A possibly qualified reference expression.
```
What's a reference expression?

ilya-biryukov replied:
Corresponds to clang's `DeclRefExpr`

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+  static bool classof(const Node *N) {
+    return N->kind() == NodeKind::RangeBasedForStmt;
+  }
+  NodeKind kind() const { return NodeKind::RangeBasedForStmt; }
```
These nodes should have accessor APIs to get significant parts of the expression, right?

ilya-biryukov replied:
Yes, some accessors are missing right now.
I'm actually considering to use TableGen to generate accessors, factory functions, etc.

gribozavr commented on clang/include/clang/Tooling/Syntax/Syntax.h:
```diff
@@ -0,0 +1,373 @@
+
+  const TokenizedBuffer &mainFile() const;
+
+  /// Get a text for a continuous range of tokens.
```
Returns text that corresponds to ...

However, it is unclear what happens to trivia -- like whitespace and comments.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+
+class TokenizedBuffer;
+
+/// A token coming directly from a file and not a macro expansion.
```
A token that came directly from a file (either main file or an included file), not from a macro expansion.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+class TokenizedBuffer;
+
+/// A token coming directly from a file and not a macro expansion.
+struct FileToken {
```
"FileTokenLocation"?

The name is similar to "Token", but this type is very unlike the "Token" type, in that it does not have detailed information about the token.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+public:
+  /// The tokens obtained after expansion.
+  llvm::ArrayRef<Token> expansionTokens(const TokenizedBuffer &B) const;
+  /// These cover the name and arguments of a macro (if any), incluging the
```
"Tokens of the macro invocation, including the macro name (always), parentheses and arguments (if any)."

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  llvm::ArrayRef<Token> expansionTokens(const TokenizedBuffer &B) const;
+  /// These cover the name and arguments of a macro (if any), incluging the
+  /// parentheses around macro arguments.
+  llvm::ArrayRef<FileToken> macroTokens(const TokenizedBuffer &B) const;
```
`invocationTokens` or `macroInvocationTokens` (but if we use the word "macro" here, then `expansionTokens` => `macroExpansionTokens` as well)

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  /// parentheses around macro arguments.
+  llvm::ArrayRef<FileToken> macroTokens(const TokenizedBuffer &B) const;
+  /// Range covering the name and arguments of a macro expansion.
+  std::pair<unsigned, unsigned> range(const TokenizedBuffer &B) const;
```
`invocationSourceRange` or `macroInvocationSourceRange` depending on what you choose for the function above.

gribozavr replied:
I'm not sure why `FileToken` and `MacroExpansion` are not using `SourceLocation`?

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  unsigned BeginExpansionToken = 0;
+  unsigned EndExpansionToken = 0;
+  unsigned BeginFileToken = 0;
+  unsigned EndFileToken = 0;
```
Please add brief doc comments to these variables.  Having read the public API of this class, I still don't have an idea what these variables are.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+public:
+  TokenizedBuffer() = default;
+  TokenizedBuffer(TokenizedBuffer &&) = default;
+  // Delete the copy constructor to avoid invalidating pointers.
```
Also delete the copy-assignment operator, and default the move-assignment operator.

And I don't think you need to explain this in a comment, it is obvious that you want to make `TokenizedBuffer` a move-only type.  Comment about "invalidating pointers" is too unspecific to be helpful.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  // Delete the copy constructor to avoid invalidating pointers.
+  TokenizedBuffer(const TokenizedBuffer &) = delete;
+  // Creates buffer from a list of tokens, assuming no macro expansions have
+  // taken place.
```
Three slashes for doc comments.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  std::vector<Token> Tokens;
+  /// Tokens forming top-level macro expansions, i.e. all macro names and macro
+  /// arguments.
+  std::vector<FileToken> ExpansionTokens;
```
I'd suggest to use the word "invocations".

"Expansion" is ambiguous: is it the call itself, or the tokens that it expanded to?

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+};
+
+/// Collects a tokenized buffer for the main file while running the frontend
+/// action. An instance of this object should be created on
```
I would suggest to not explain cross-references in comments -- they become stale real quick.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+/// running FrontendAction::Execute().
+class TokenCollector {
+public:
+  /// Adds the hooks to collect the tokens. Should be called before the
```
Adds => Registers

before the => before

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  /// CreateASTConsumer().
+  TokenCollector(Preprocessor &P);
+
+  /// Consumes the result. Should be called after preprocessing is finished,
```
"Finalizes token collection"

Of course a function called "consume" consumes the result :)

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+public:
+  /// Adds the hooks to collect the tokens. Should be called before the
+  /// preprocessing starts, i.e. as a part of BeginSourceFile() or
+  /// CreateASTConsumer().
```
Qualify methods with class names?

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  TokenCollector(Preprocessor &P);
+
+  /// Consumes the result. Should be called after preprocessing is finished,
+  /// i.e. after running Execute().
```
Qualify methods with class names?

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+
+  /// Consumes the result. Should be called after preprocessing is finished,
+  /// i.e. after running Execute().
+  TokenizedBuffer consume() &&;
```
"finalize()"?

gribozavr replied:
Use `LLVM_NODISCARD`.

gribozavr commented on clang/lib/Tooling/Syntax/TokenizedBuffer.cpp:
```diff
@@ -0,0 +1,357 @@
+llvm::Optional<std::pair<unsigned, unsigned>>
+TokenizedBuffer::toOffsetRange(const Token *Begin, const Token *End,
+                               const SourceManager &SM) const {
+  assert(Begin < End);
```
Also assert that Begin and End point within `Tokens`.

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  llvm::ArrayRef<MacroExpansion> expansions() const { return Expansions; }
+  /// File tokens that formed MacroExpansions. Most users don't need this.
+  llvm::ArrayRef<FileToken> expansionTokens() const { return ExpansionTokens; }
+  /// Attempt the map a range of expanded tokens into a continuous range of
```
"Finds a continuous range in the source file that corresponds to a range of tokens produced by preprocessor.  Input and output ranges are half-open.

It might not be possible to find the corresponding text in the source file if the tokens cross macro expansion boundaries."

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  llvm::ArrayRef<FileToken> expansionTokens() const { return ExpansionTokens; }
+  /// Attempt the map a range of expanded tokens into a continuous range of
+  /// offsets inside the original source file. The tranformation may not be
+  /// possible if the range requires changing the macro expansions.
```
I did not understand what "changing the macro expansions" means in this context since this operation is not changing anything...

Maybe a couple of examples would help, like, this kind of operation is supported, and this will fail.

Given this source file:

```
#define FIRST f1 f2 f3
#define SECOND s1 s2 3

a FIRST b SECOND c  // expansion: a f1 f2 f3 b s1 s2 s3 c
```

`toOffsetRange` will map tokens like this:

```
input range => output range
a => a
s1 s2 s3 => SECOND
a f1 f2 f3 => a FIRST
a f1 => can't map
s1 s2 => ???
```

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  llvm::Optional<std::pair<unsigned, unsigned>>
+  toOffsetRange(const Token *Begin, const Token *End,
+                const SourceManager &SM) const;
+  /// Find a macro call by its firts expanded token.
```
firts => first

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  toOffsetRange(const Token *Begin, const Token *End,
+                const SourceManager &SM) const;
+  /// Find a macro call by its firts expanded token.
+  const MacroExpansion *findMacroCall(const Token *Begin) const;
```
"findMacroInvocation"

gribozavr commented on clang/include/clang/Tooling/Syntax/TokenizedBuffer.h:
```diff
@@ -0,0 +1,113 @@
+  // taken place.
+  TokenizedBuffer(std::vector<clang::Token> Tokens);
+
+  /// All tokens from the result of preprocessor expansion, i.e. the list of
```
"All tokens produced by the preprocessor (incl. tokens from included headers).

Source locations in the Clang AST always point into one of these tokens."

gribozavr commented on clang/lib/Tooling/Syntax/TokenizedBuffer.cpp:
```diff
@@ -0,0 +1,357 @@
+    FileID File;
+    unsigned Offset;
+    std::tie(File, Offset) = SM.getDecomposedLoc(L);
+    if (File != ExpansionFile || Offset <= ExpansionEndOffset)
```
I don't quite understand when this can happen.  Is it defensive coding, a tricky unhandled case, or intentional logic?

ilya-biryukov replied:
I think this can actually happen during tentative parsing.
I'll add the corresponding tests into the phabricator review and double-check that this case is covered.

ilya-biryukov committed:
[Testing] Move clangd::Annotations to llvm testing support

Summary:
Annotations allow writing nice-looking unit test code when one needs
access to locations from the source code, e.g. running code completion
at particular offsets in a file. See comments in Annotations.cpp for
more details on the API.

Also got rid of a duplicate annotations parsing code in clang's code
complete tests.

Reviewers: gribozavr

Subscribers: mgorny, hiraditya, ioeric, MaskRay, jkorous, arphaman, kadircet, jdoerfert, cfe-commits, llvm-commits

Tags: #clang, #llvm

Differential Revision: https://reviews.llvm.org/D59814

ilya-biryukov committed:
[Syntax] Introduce mutable syntax trees

This commit collects everything not related to TokenBuffers.
It needs to be split into smaller pieces to prepare it for review.

ilya-biryukov committed:
[Syntax] Add a findMacroCall helper

ilya-biryukov commented:
Accidentally pushed "ready for review", sorry.
